File: src/routes/customers.ts
import { Router, Request, Response } from "express";
import type { CustomerService } from "../services/customer.service.js";
import { asyncHandler } from "../middleware/asyncHandler.js";
import { paginationSchema } from "../schemas/index.js";

export function createCustomerRouter(service: CustomerService): Router {
  const router = Router();

  router.get("/", asyncHandler(async (req: Request, res: Response) => {
    const pagination = paginationSchema.parse(req.query);
    res.json(await service.listCustomers(pagination));
  }));

  router.post("/", asyncHandler(async (req: Request, res: Response) => {
    const data = await service.createCustomer(req.body);
    res.status(201).json(data);
  }));

  return router;
}

File: app/routers/customers.py
from fastapi import APIRouter, Depends, Query

from app.middleware.auth import CurrentUser, get_current_user
from app.repositories.customer import CustomerRepository
from app.schemas import CustomerInput
from app.services.customer import CustomerService

repo = CustomerRepository()
service = CustomerService(repo)
router = APIRouter(prefix="/api/customers", tags=["customers"])


@router.get("/")
def list_customers(
    limit: int = Query(default=50, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
    user: CurrentUser = Depends(get_current_user),
):
    return service.list_customers(limit, offset)


@router.post("/", status_code=201)
def create_customer(body: CustomerInput, user: CurrentUser = Depends(get_current_user)):
    return service.create_customer(body.model_dump())
